app/frontend/src/cljs/teet/asset/cost_items_view.cljs

```diff
@@ -262,111 +262,121 @@
+                        :asset-schema/keys [unit] :as attr} attrs
+                       :let [type (:db/ident valueType)
+                             unit (if (= ident :component/quantity)
+                                    (:component/quantity-unit ctype)
+                                    unit)]]
+                   ^{:key (str ident)}
+                   [attribute-grid-item
+                    [form/field {:attribute ident
+                                 :required? mandatory?
+                                 :validate (r/partial validate
+                                                      (:db/ident valueType)
+                                                      component-oid
+                                                      cost-item-data
+                                                      (select-keys attr
+                                                                   [:attribute/min-value
+                                                                    :attribute/max-value
+                                                                    :attribute/min-value-ref
+                                                                    :attribute/max-value-ref]))}
+                     (cond
+                       ;; Selection value
+                       (= type :db.type/ref)
+                       (if locked?
+                         [display-list-item {:label (label attr)
+                                             :rotl rotl}]
+                         [select/form-select
+                          {:id ident
+                           :label (label attr)
+                           :show-empty-selection? true
+                           :items (mapv :db/ident (:enum/_attribute attr))
+                           :format-item (comp label rotl)
+                           ;; Show error message as tooltip instead of adding a span, so that input
+                           ;; elements in the grid will stay aligned.
+                           :error-tooltip? true}])
+
+                       (= type :db.type/instant)
+                       (if locked?
+                         [display-input {:label (label attr)
+                                         :format fmt/date}]
+                         [date-picker/date-input {:label (label attr)
+                                                  :error-tooltip? true}])
+
+                       ;; Text field
+                       :else
+                       (if locked?
+                         [display-input {:label (label attr)
+                                         :unit unit}]
+                         [text-field/TextField
+                          {:label (label attr)
+                           :read-only? locked?
+                           :end-icon (when unit
+                                       (text-field/unit-end-icon unit))
+                           :error-tooltip? true}]))]]))]]))]))})))
```

Comment:
Loks like too long function to me, could use a bit of refactoring into smaller ones.